// @tanstack/vue-query/build/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var queryCore = require('@tanstack/query-core');
var useQueryClient = require('./useQueryClient.js');
var vueQueryPlugin = require('./vueQueryPlugin.js');
var queryClient = require('./queryClient.js');
var queryCache = require('./queryCache.js');
var mutationCache = require('./mutationCache.js');
var useQuery = require('./useQuery.js');
var useQueries = require('./useQueries.js');
var useInfiniteQuery = require('./useInfiniteQuery.js');
var useMutation = require('./useMutation.js');
var useIsFetching = require('./useIsFetching.js');
var useIsMutating = require('./useIsMutating.js');
var utils = require('./utils.js');



exports.useQueryClient = useQueryClient.useQueryClient;
exports.VueQueryPlugin = vueQueryPlugin.VueQueryPlugin;
exports.QueryClient = queryClient.QueryClient;
exports.QueryCache = queryCache.QueryCache;
exports.MutationCache = mutationCache.MutationCache;
exports.useQuery = useQuery.useQuery;
exports.useQueries = useQueries.useQueries;
exports.useInfiniteQuery = useInfiniteQuery.useInfiniteQuery;
exports.useMutation = useMutation.useMutation;
exports.useIsFetching = useIsFetching.useIsFetching;
exports.useIsMutating = useIsMutating.useIsMutating;
exports.VUE_QUERY_CLIENT = utils.VUE_QUERY_CLIENT;
Object.keys(queryCore).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return queryCore[k]; }
  });
});
