// @tanstack/vue-query/build/lib/useBaseQuery.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vueDemi = require('vue-demi');
var useQueryClient = require('./useQueryClient.js');
var utils = require('./utils.js');

function useBaseQuery(Observer, arg1, arg2 = {}, arg3 = {}) {
  var _options$value$queryC;

  const options = vueDemi.computed(() => parseQueryArgs(arg1, arg2, arg3));
  const queryClient = (_options$value$queryC = options.value.queryClient) != null ? _options$value$queryC : useQueryClient.useQueryClient(options.value.queryClientKey);
  const defaultedOptions = vueDemi.computed(() => {
    const defaulted = queryClient.defaultQueryOptions(options.value);
    defaulted._optimisticResults = queryClient.isRestoring.value ? 'isRestoring' : 'optimistic';
    return defaulted;
  });
  const observer = new Observer(queryClient, defaultedOptions.value);
  const state = vueDemi.reactive(observer.getCurrentResult());
  const unsubscribe = vueDemi.ref(() => {// noop
  });
  vueDemi.watch(queryClient.isRestoring, isRestoring => {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!isRestoring) {
      unsubscribe.value();
      unsubscribe.value = observer.subscribe(result => {
        utils.updateState(state, result);
      });
    }
  }, {
    immediate: true
  });
  vueDemi.watch(defaultedOptions, () => {
    observer.setOptions(defaultedOptions.value);
    utils.updateState(state, observer.getCurrentResult());
  }, {
    deep: true
  });
  vueDemi.onScopeDispose(() => {
    unsubscribe.value();
  });

  const suspense = () => {
    return new Promise(resolve => {
      let stopWatch = () => {//noop
      };

      const run = () => {
        if (defaultedOptions.value.enabled !== false) {
          const optimisticResult = observer.getOptimisticResult(defaultedOptions.value);

          if (optimisticResult.isStale) {
            stopWatch();
            resolve(observer.fetchOptimistic(defaultedOptions.value));
          } else {
            stopWatch();
            resolve(optimisticResult);
          }
        }
      };

      run();
      stopWatch = vueDemi.watch(defaultedOptions, run, {
        deep: true
      });
    });
  };

  return { ...vueDemi.toRefs(vueDemi.readonly(state)),
    suspense
  };
}
function parseQueryArgs(arg1, arg2 = {}, arg3 = {}) {
  const plainArg1 = vueDemi.unref(arg1);
  const plainArg2 = vueDemi.unref(arg2);
  const plainArg3 = vueDemi.unref(arg3);
  let options = plainArg1;

  if (!utils.isQueryKey(plainArg1)) {
    options = plainArg1;
  } else if (typeof plainArg2 === 'function') {
    options = { ...plainArg3,
      queryKey: plainArg1,
      queryFn: plainArg2
    };
  } else {
    options = { ...plainArg2,
      queryKey: plainArg1
    };
  }

  return utils.cloneDeepUnref(options);
}

exports.parseQueryArgs = parseQueryArgs;
exports.useBaseQuery = useBaseQuery;
