// @tanstack/vue-query/build/lib/useMutation.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vueDemi = require('vue-demi');
var queryCore = require('@tanstack/query-core');
var utils = require('./utils.js');
var useQueryClient = require('./useQueryClient.js');

function useMutation(arg1, arg2, arg3) {
  var _options$value$queryC;

  const options = vueDemi.computed(() => {
    return parseMutationArgs(arg1, arg2, arg3);
  });
  const queryClient = (_options$value$queryC = options.value.queryClient) != null ? _options$value$queryC : useQueryClient.useQueryClient(options.value.queryClientKey);
  const observer = new queryCore.MutationObserver(queryClient, queryClient.defaultMutationOptions(options.value));
  const state = vueDemi.reactive(observer.getCurrentResult());
  const unsubscribe = observer.subscribe(result => {
    utils.updateState(state, result);
  });

  const mutate = (variables, mutateOptions) => {
    observer.mutate(variables, mutateOptions).catch(() => {// This is intentional
    });
  };

  vueDemi.watch(options, () => {
    observer.setOptions(queryClient.defaultMutationOptions(options.value));
  }, {
    deep: true
  });
  vueDemi.onScopeDispose(() => {
    unsubscribe();
  });
  const resultRefs = vueDemi.toRefs(vueDemi.readonly(state));
  return { ...resultRefs,
    mutate,
    mutateAsync: state.mutate,
    reset: state.reset
  };
}
function parseMutationArgs(arg1, arg2, arg3) {
  const plainArg1 = vueDemi.unref(arg1);
  const plainArg2 = vueDemi.unref(arg2);
  let options = plainArg1;

  if (utils.isMutationKey(plainArg1)) {
    if (typeof plainArg2 === 'function') {
      const plainArg3 = vueDemi.unref(arg3);
      options = { ...plainArg3,
        mutationKey: plainArg1,
        mutationFn: plainArg2
      };
    } else {
      options = { ...plainArg2,
        mutationKey: plainArg1
      };
    }
  }

  if (typeof plainArg1 === 'function') {
    options = { ...plainArg2,
      mutationFn: plainArg1
    };
  }

  return utils.cloneDeepUnref(options);
}

exports.parseMutationArgs = parseMutationArgs;
exports.useMutation = useMutation;
